datalad/core/local/diff.py

```diff
@@ -0,0 +1,335 @@
+                        "the reference dataset %s: %s",
+                        ds, str_path),
+                    logger=lgr,
+                )
+                continue
+
+            ps.append(p)
+        path = ps
+
+    # TODO we might want to move away from the single-pass+immediate-yield
+    # paradigm for this command. If we gather all information first, we
+    # could do post-processing and detect when a file (same gitsha, or same
+    # key) was copied/moved from another dataset. Another command (e.g.
+    # rev-save) could act on this information and also move/copy
+    # availability information or at least enhance the respective commit
+    # message with cross-dataset provenance info
+
+    # cache to help avoid duplicate status queries
+    content_info_cache = {}
+    for res in _diff_ds(
+            ds,
+            fr,
+            to,
+            constant_refs,
+            recursion_limit
+            if recursion_limit is not None and recursive
+            else -1 if recursive else 0,
+            # TODO recode paths to repo path reference
+            origpaths=None if not path else OrderedDict(path),
+            untracked=untracked,
+            annexinfo=annex,
+            eval_file_type=True,
+            cache=content_info_cache):
+        res.update(
+            refds=ds.path,
+            logger=lgr,
+            action='diff',
+        )
+        yield res
+
+
+def _diff_ds(ds, fr, to, constant_refs, recursion_level, origpaths, untracked,
+             annexinfo, eval_file_type, cache):
+    if not ds.repo:
+        # asked to query a subdataset that is not available
+        lgr.debug("Skip diff of unavailable subdataset: %s", ds)
+        return
+
+    repo_path = ds.repo.pathobj
+    # filter and normalize paths that match this dataset before passing them
+    # onto the low-level query method
+    paths = None if origpaths is None \
+        else OrderedDict(
+            (repo_path / p.relative_to(ds.pathobj), goinside)
+            for p, goinside in iteritems(origpaths)
+            if ds.pathobj in p.parents or (p == ds.pathobj and goinside)
+        )
+    try:
+        lgr.debug("diff %s from '%s' to '%s'", ds, fr, to)
+        diff_state = ds.repo.diffstatus(
+            fr,
+            to,
+            paths=None if not paths else [p for p in paths],
+            untracked=untracked,
+            eval_file_type=eval_file_type,
+            _cache=cache)
+    except ValueError as e:
+        msg_tmpl = "reference '{}' invalid"
+        # not looking for a debug repr of the exception, just the message
+        estr = text_type(e)
+        if msg_tmpl.format(fr) in estr or msg_tmpl.format(to) in estr:
+            yield dict(
+                path=ds.path,
+                status='impossible',
+                message=estr,
+            )
+            return
+
+    if annexinfo and hasattr(ds.repo, 'get_content_annexinfo'):
```

Comment:
why is it not just 
```suggestion
    if annexinfo and isinstance(ds.repo, AnnexRepo):
```
? (the same check exists also in datalad/core/local/status.py)  Is it left for some compatibility with pre-0.12rc datalad which did not have it?

Reply:
There was a time, when AnnexRepo was kept intentionally uneducated ;-)

Reply:
I thought about this, and I want to keep it as-is. The test checks exactly what is needed: an object with this method. It is good as is, and changing it only has the potential to be stricter than necessary.

Reply:
Should we move all code to similar duck typing for uniformity? ;-) but whatever, let's call it stylistic and use interchangeably